src/main/java/org/folio/auth/authtokenmodule/tokens/Token.java

```diff
@@ -251,30 +257,56 @@ protected void validateCommon(HttpServerRequest request) throws TokenValidationE
       }
 
       if (request == null) {
-        return;
+        promise.complete(this);
+        return promise.future();
       }
 
       // Check that some items in the headers match what are in the token.
-      String headerTenant = request.headers().get(XOkapiHeaders.TENANT);
-      if (!claims.getString("tenant").equals(headerTenant)) {
-        throw new TokenValidationException("Tenant mismatch: tenant in header does not equal tenant in token", 403);
-      }
       String headerUserId = request.headers().get(XOkapiHeaders.USER_ID);
       String claimsUserId = claims.getString("user_id");
       if (headerUserId != null && claimsUserId != null && !claimsUserId.equals(headerUserId)) {
         throw new TokenValidationException("User id in header does not equal user id in token", 403);
       }
+
+      String headerTenant = request.headers().get(XOkapiHeaders.TENANT);
+      if (!claims.getString("tenant").equals(headerTenant)) {
+        isCrossTenantRequest(context)
+          .compose(aResult -> Boolean.TRUE.equals(aResult) ? Future.succeededFuture(this)
+            : Future.failedFuture(new TokenValidationException(TENANT_MISMATCH_EXCEPTION_MESSAGE, 403)))
+          .onComplete(promise);
+      } else {
+        promise.complete(this);
+      }
     } catch (TokenValidationException e) {
-      throw e;
+      promise.fail(e);
     } catch (Exception e) {
       logger.error(e.getMessage(), e);
-      throw new TokenValidationException("Unexpected token validation exception", e, 500);
+      promise.fail(new TokenValidationException("Unexpected token validation exception", e, 500));
     }
+
+    return promise.future();
   }
 
   protected boolean tokenIsExpired() {
     Long nowTime = Instant.now().getEpochSecond();
     Long expiration = claims.getLong("exp");
     return nowTime > expiration;
   }
+
+  protected Future<Boolean> isCrossTenantRequest(TokenValidationContext context) {
```

Comment:
please add a comment to this method. It is not clear why isUserTenantNotEmpty signal that the request is a cross tenant request.

Reply:
Agree, but it would be preferable to add javadocs to all the new methods (following the practice of other method in  this module).

Reply:
Done